src/libasr/pass/pass_utils.cpp

```diff
@@ -646,50 +646,78 @@ namespace LFortran {
                         increment = - ASR::down_cast<ASR::IntegerConstant_t>(u->m_arg)->m_n;
                     } else {
                         // This is the case when increment operator is not a
-                        // constant, and so we need an if statement to check that
+                        // constant, and so we need some conditions to check
                         // in the backend and generate while loop according
                         // to avoid infinite loops.
                         not_constant_inc = true;
                     }
+
                     if (not_constant_inc) {
-                        ASR::ttype_t *int_type = LFortran::ASRUtils::TYPE(ASR::make_Integer_t(al,
-                                    loc, 4, nullptr, 0));
+                        ASR::expr_t *target = loop.m_head.m_v;
+                        int a_kind = ASRUtils::extract_kind_from_ttype_t(ASRUtils::expr_type(target));
+                        ASR::ttype_t *int_type = LFortran::ASRUtils::TYPE(ASR::make_Integer_t(al, loc, a_kind, nullptr, 0));
+
                         ASR::ttype_t *log_type = ASRUtils::TYPE(
                             ASR::make_Logical_t(al, loc, 1, nullptr, 0));
```

Comment:
We use `logical(4)` as the default logical:
```suggestion
                            ASR::make_Logical_t(al, loc, 4, nullptr, 0));
```
That will cleanup the diffs in the tests as well.